devito/interfaces.py

```diff
@@ -440,44 +445,126 @@ class DenseData(TensorData):
+                setattr(self.__class__, 'd%sr' % dim.name,
+                        property(dxr, 'Symbol for the first derivative '
+                                 'wrt %s with a right stencil' % dim.name))
+
+                # Second derivative
+                dx2 = partial(generic_derivative, order=2, dim=dim,
+                              width=self.space_order / 2)
+                setattr(self.__class__, 'd%s2' % dim.name,
+                        property(dx2, 'Symbol for the second derivative '
+                                 'wrt the %s dimension' % dim.name))
+
+                # Fourth derivative
+                dx4 = partial(generic_derivative, order=4, dim=dim,
+                              width=max(int(self.space_order / 2), 2))
+                setattr(self.__class__, 'd%s4' % dim.name,
+                        property(dx4, 'Symbol for the fourth derivative '
+                                 'wrt the %s dimension' % dim.name))
+
+                for dim2 in self.space_dimensions:
+                    # First cross derivative
+                    dxy = partial(cross_derivative, order=self.space_order,
+                                  dims=(dim, dim2))
+                    setattr(self.__class__, 'd%s%s' % (dim.name, dim2.name),
+                            property(dxy, 'Symbol for the cross derivative '
+                                     'wrt the %s and %s dimension' %
+                                     (dim.name, dim2.name)))
+
+                    # Second cross derivative
+                    dx2y2 = partial(second_cross_derivative, dims=(dim, dim2),
+                                    order=self.space_order)
+                    setattr(self.__class__, 'd%s2%s2' % (dim.name, dim2.name),
+                            property(dx2y2, 'Symbol for the cross derivative '
+                                     'wrt the %s and %s dimension' %
+                                     (dim.name, dim2.name)))
+
     @classmethod
     def _indices(cls, **kwargs):
         """Return the default dimension indices for a given data shape
 
+        :param grid: :class:`Grid` that defines the spatial domain.
         :param dimensions: Optional, list of :class:`Dimension`
                            objects that defines data layout.
-        :param shape: Optional, shape of the spatial data to
-                      automatically infer dimension symbols.
         :return: Dimension indices used for each axis.
+
+        ..note::
+
+        Only one of :param grid: or :param dimensions: is required.
         """
+
+        grid = kwargs.get('grid', None)
         dimensions = kwargs.get('dimensions', None)
-        if dimensions is None:
-            # Infer dimensions from default and data shape
-            if 'shape' not in kwargs:
-                error("Creating symbolic data objects requries either"
-                      "a 'shape' or 'dimensions' argument")
+        if grid is None:
```

Comment:
nitpicking: move ``dimensions = kwargs.get(...)`` inside this first ``if``

Reply:
No, it's used in the else branch too to warn about implicit overrides.